### PR TITLE
WT-12604 Create tool to check for bit flip errors

### PR DESCRIPTION
**Pull request by keitharnoldsmith**

### CODE REVIEW

File: src/docs/wtperf.dox [hunk 1]
```diff
@@ -128,6 +128,8 @@ configuration options:
 DO NOT EDIT: THIS PART OF THE FILE IS GENERATED BY dist/s_docs.
 \endif
 
+@par backup_complete (unsigned int, default=0)
+when doing backups, either just perform source-side reading or take a complete backup
 @par backup_interval (unsigned int, default=0)
```
Comment (keitharnoldsmith) [inline src/docs/wtperf.dox:133]: This is unrelated to my change. It came from running `s_docs` (via `s_all`).

Comment (sueloverso) [reply src/docs/wtperf.dox:133]: That's from my recent merge. But `s_docs` or `s_all` did not fail on my PR.

File: tools/checksum_bitflip/checksum_bitflip.c [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+/*
+ * This program takes a file and a checksum and checks whether the checksum
+ * matches the contents of the file, or if any checksum produced by flipping
+ * a single bit of the file contents matches.
+ *
+ * This intended for use when debugging WT checksum mismatches when we suspect
```
Comment (sueloverso) [inline tools/checksum_bitflip/checksum_bitflip.c:38]: "This is intended..."

File: tools/checksum_bitflip/checksum_bitflip.c [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+main(int argc, char *argv[])
+{
+    char *buffer, *filename;
+    uint8_t mask;
+    int bit, fd;
+    size_t byte, size;
+    ssize_t io_bytes;
+    uint32_t cksum_target;
+    int64_t tmp;
+    struct stat statbuf;
```
Comment (sueloverso) [inline tools/checksum_bitflip/checksum_bitflip.c:57]: Reorder by KNF decreasing size.

File: tools/checksum_bitflip/checksum_bitflip.c [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    tmp = strtol(argv[1], NULL, 16);
+    if ((tmp ^ (tmp & 0xFFFFFFFF)) != 0) {
+        fprintf (stderr, "Target checksum must be 32-bits\n");
+        exit(1);
+    }
+    cksum_target = (uint32_t) tmp;
+
+    filename = argv[2];
+    fd = open(filename, O_RDONLY);
```
Comment (sueloverso) [inline tools/checksum_bitflip/checksum_bitflip.c:72]: Error checking here?

File: tools/checksum_bitflip/checksum_bitflip.c [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    io_bytes = read(fd, buffer, size);   
+    if ((size_t) io_bytes != size) {
+        printf ("Read of %s returned %zd\n", filename, io_bytes);
+        exit(1);
+    }
+
+    /* See if the checksum matches the file contents. */
+    if (__wt_checksum_sw(buffer, size) == cksum_target) {
+        printf ("Checksum match without flipping bits\n");
```
Comment (sueloverso) [inline tools/checksum_bitflip/checksum_bitflip.c:85]: Either "Checksums match" or "Checksum matches"
Nit: no space between `printf` and the opening `(`. There are a bunch in the file, I won't call them all out but please fix them all.

File: tools/checksum_bitflip/checksum_bitflip.c [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    /*
+     * Iterate through the file contents flipping individual bits and checking
+     * whether the resulting data generates a matching checksum.
+     */
+    for (byte = 0; byte < size; byte++) {
+        mask = 0x01;
+        for (bit = 0; bit < 8; bit++) {
+            buffer[byte] ^= mask;
+            if (__wt_checksum_sw(buffer, size) == cksum_target) {
+                printf ("Checksum match when flipping bit %d of byte %zu\n", bit, byte);
```
Comment (sueloverso) [inline tools/checksum_bitflip/checksum_bitflip.c:98]: Same grammar change needed here too.

File: tools/checksum_bitflip/checksum_bitflip.c [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                exit(0);
+            }
+            buffer[byte] ^= mask;
+            mask <<= 1;
+        }
+    }
+    printf ("No checksum match\n");
```
Comment (sueloverso) [inline tools/checksum_bitflip/checksum_bitflip.c:105]: Would it be helpful here and/or above to say something like "This may indicate a hardware issue." or "This may not..." so that it is clear what the result points to?

File: tools/checksum_bitflip/checksum_bitflip.c [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    }
+    cksum_target = (uint32_t) tmp;
+
+    filename = argv[2];
+    fd = open(filename, O_RDONLY);
+    fstat(fd, &statbuf);
+    size = (size_t) statbuf.st_size;
+    buffer = dmalloc(size);
+
+    io_bytes = read(fd, buffer, size);
```
Comment (sueloverso) [inline tools/checksum_bitflip/checksum_bitflip.c:77]: Error checking?

File: tools/checksum_bitflip/checksum_bitflip.c [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        printf ("Usage: %s checksum filename\n", argv[0]);
+        exit(1);
+    }
+
+    tmp = strtol(argv[1], NULL, 16);
+    if ((tmp ^ (tmp & 0xFFFFFFFF)) != 0) {
+        fprintf (stderr, "Target checksum must be 32-bits\n");
+        exit(1);
+    }
+    cksum_target = (uint32_t) tmp;
```
Comment (sueloverso) [inline tools/checksum_bitflip/checksum_bitflip.c:69]: No space between cast and `tmp`.

File: tools/checksum_bitflip/checksum_bitflip.c [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    fd = open(filename, O_RDONLY);
+    fstat(fd, &statbuf);
+    size = (size_t) statbuf.st_size;
+    buffer = dmalloc(size);
+
+    io_bytes = read(fd, buffer, size);   
+    if ((size_t) io_bytes != size) {
+        printf ("Read of %s returned %zd\n", filename, io_bytes);
+        exit(1);
+    }
```
Comment (sueloverso) [inline tools/checksum_bitflip/checksum_bitflip.c:81]: Hmm, I thought WT checksummed each block/page it writes, not the entire file. So I thought this would be checksumming 32K chunks or whatever size.

**Review by sueloverso — COMMENTED** [review]
I'll take another look when done.

File: tools/checksum_bitflip/checksum_bitflip.c (hunk 2 above)
Comment (keitharnoldsmith) [reply tools/checksum_bitflip/checksum_bitflip.c:38]: Done

File: tools/checksum_bitflip/checksum_bitflip.c (hunk 3 above)
Comment (keitharnoldsmith) [reply tools/checksum_bitflip/checksum_bitflip.c:57]: Done

File: tools/checksum_bitflip/checksum_bitflip.c (hunk 6 above)
Comment (keitharnoldsmith) [reply tools/checksum_bitflip/checksum_bitflip.c:98]: Done

File: tools/checksum_bitflip/checksum_bitflip.c (hunk 4 above)
Comment (keitharnoldsmith) [reply tools/checksum_bitflip/checksum_bitflip.c:72]: Yup. I meant to do that. But obviously missed it. I've updated the code to assert the system calls work.

File: tools/checksum_bitflip/checksum_bitflip.c (hunk 5 above)
Comment (keitharnoldsmith) [reply tools/checksum_bitflip/checksum_bitflip.c:85]: Done. Also updated the `exit` calls to avoid that space. You can tell this is a place where my native C style doesn't match the WT style :-).

File: tools/checksum_bitflip/checksum_bitflip.c (hunk 7 above)
Comment (keitharnoldsmith) [reply tools/checksum_bitflip/checksum_bitflip.c:105]: I'd rather not have the output of the tool assume the use case. Maybe in the future people would want to use this for other reasons.

File: tools/checksum_bitflip/checksum_bitflip.c (hunk 8 above)
Comment (keitharnoldsmith) [reply tools/checksum_bitflip/checksum_bitflip.c:77]: Done.

File: tools/checksum_bitflip/checksum_bitflip.c (hunk 9 above)
Comment (keitharnoldsmith) [reply tools/checksum_bitflip/checksum_bitflip.c:69]: Done (and other casts).

File: tools/checksum_bitflip/checksum_bitflip.c (hunk 10 above)
Comment (keitharnoldsmith) [reply tools/checksum_bitflip/checksum_bitflip.c:81]: Ah. I wasn't clear about this. I don't point this at a `*.wt` file. Instead I take the hex dump we generate on checksum mismatch, and convert that into binary. Then run this tool on the resulting file. The advantage is that you can do that on customer failures where we don't have access to the WiredTiger file.

I've added text to the opening comment explaining this.

But I like the idea of being able to specify an offset and size as optional parameters so you could also use this on a WiredTiger file. I'll create a follow-up ticket for that.

Comment (keitharnoldsmith) [reply tools/checksum_bitflip/checksum_bitflip.c:81]: Meta-comment. Some of the issues here are because we apparently don't run clang-format on things in the `tools/` directory. It looks like this is the first C program we've added there.

**Comment by keitharnoldsmith** [comment]
Meta-comment. Some of the issues @sueloverso identified would have been handled by clang-format. But apparently we don't run that on field in the tools/ directory. It looks like this is the first C program we've added there.

**Review by sueloverso — APPROVED** [review]
LGTM, I have one additional small comment.

File: tools/checksum_bitflip/checksum_bitflip.c [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+ * This is intended for use when debugging WT checksum mismatches when we suspect
+ * that the issue is the result of faulty hardware causing bit flips in memory.
+ *
+ * Usage:
+ *     checksum_bitflip cksum filename
+ *
+ * The checksum argument should be a hex string. The file should contain the data
+ * to be tested. Typical usage is to generate this data from the block dump printed
+ * when WiredTiger reports a checksum mismatch, for example by copying the hex
+ * strings into a file and running "xxd -p -r" on it.
```
[the remark this answers is not in the document]
Comment (keitharnoldsmith) [reply tools/checksum_bitflip/checksum_bitflip.c:47]: Good point. I've updated the other scripts. I've also narrowed the scope to only run in the `tools/checkpoint_bitflip` directory, rather than the entire `tools/` directory. 

It's a fair question whether we should run these scripts across the rest of the tools, but if we did, there are a bunch of strings and other things that would need to be cleaned up, and that doesn't belong in this ticket. I'll create a separate ticket and where we can decide whether and how much we want this checking on the other tools.
